scripts/core/editor3/reducers/editor3.tsx

```diff
@@ -262,6 +269,46 @@ function escapeRegExp(string) {
     return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
 }
 
+function replaceText(
+    editorState: EditorState,
+    selection: SelectionState,
+    text: string,
+) {
+    const newContent = Modifier.replaceText(
+        editorState.getCurrentContent(),
+        selection,
+        text,
+    );
+
+    return EditorState.push(editorState, newContent, 'insert-characters');
+}
+
+function insertTab(state: EditorState, selection: SelectionState, block: ContentBlock, tabString: string) {
+    const indentInsteadOfInsert = !selection.isCollapsed();
+    const selectionForInsert = indentInsteadOfInsert
```

Comment:
Why aren't you using the original selection?

Reply:
I do use it, but I change it when the selection is collapsed so instead of inserting a tab there I indent the text. Draft-js doesn't support text indentation yet so I just select the first character in the block, and I insert a tab there.

Reply:
How about inserting the tab at the cursor instead of at the first character and leave it up to the user to jump to the beginning of the line beforehand if that's where he wants the tab. That's how it works in most editors. You may avoid having to modify and restore selection this way too.

Reply:
We thought that when selecting text and pressing tab we could indent the text, as it is how a lot of editors work. But I guess it will be a very minor usage, the main feature (inserting tabs) works anyway, without restoring the selection

Reply:
Which editors indent the entire line when your selection is expanded? I tested it with libre office, geddit, google docs and vscode. google docs and vscode indent the line only if the entire line is selected. Otherwise, the selection gets replaced with a tab in all these editors.

> the main feature (inserting tabs) works anyway, without restoring the selection

when is restoring selection required then?

Reply:
I can do it when the entire line then, or even when you have several blocks selected I could indent those...

Restoring the selection is only required when I indent (cause I basically make a new selection at the start, insert the text, and restore). If there's no text selected the selection stays the same

Reply:
Forget it, I think I'll go back and stick to just inserting tabs, cause this is getting complicated and we'll probably have bugs for something is not really required

Reply:
Yep, I agree about keeping it simple. It's easy to get into undo/redo bugs when forcing a selection.